### PR TITLE
[inductor] Enable CSE on masked loads

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #98304
* __->__ #98303

Currently the `TritonKernel.mask_loads` context manager calls
`swap_buffers` which creates a new CSE context. So, code generated in
different mask contexts cannot be CSE'd even if their masks are the
same. This fixes the issue by not calling `swap_buffers` and instead
having `load` manually check if a `"tmp"` name appears in the mask
meaning the load needs to be generated in the compute buffer.

Currently, simple programs involving padding will result in duplcate
masked loads. e.g. the generated code for
```python
def forward():
    a = torch.nn.functional.pad(x, (0, 1))
    return a + a
```

contains the lines

```python
    tmp3 = tl.load(in_ptr0 + (x1 + tl.zeros([XBLOCK], tl.int32)), tmp2 & xmask, other=0)
    tmp4 = tl.where(tmp2, tmp3, 0.0)
    tmp5 = tl.load(in_ptr0 + (x1 + tl.zeros([XBLOCK], tl.int32)), tmp2 & xmask, other=0)
    tmp6 = tl.where(tmp2, tmp5, 0.0)
```

With this change, the duplicates are removed.

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

**Review by ngimel — APPROVED** [review]
Cool, any perf improvements from that?

**Comment by peterbell10** [comment]
Trying it out I didn't see any perf improvements so I looked into the triton IR and their CSE pass was catching this already. Oh well, at least the output's cleaner. 

There might also be some load-after-store conditions where triton can't reason that the memory regions don't overlap in which case we may gain some performance there.
